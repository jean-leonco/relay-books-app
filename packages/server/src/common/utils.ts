import { Model, Types } from 'mongoose';
import { fromGlobalId } from 'graphql-relay';

// https://stackoverflow.com/a/2593661/710693
export const escapeRegex = (str: string) => `${str}`.replace(/[.?*+^$[\]\\(){}|-]/g, '\\$&');

export const PLATFORM = {
  APP: 'APP',
  UNKNOWN: 'UNKNOWN',
};

export const getPlatform = (platform?: string | null): string => {
  if (typeof platform !== 'string') {
    return PLATFORM.UNKNOWN;
  }

  //if (platform.includes('ios') || platform.includes('android')) {
  //  return PLATFORM.APP;
  //}

  if (platform === PLATFORM.APP) {
    return PLATFORM.APP;
  }

  return PLATFORM.UNKNOWN;
};

export function kilometersToEarthRadians(kilometers: number) {
  return kilometers / 6378.1;
}

// returns an ObjectId given an param of unknown type
export const getObjectId = (target: string | Model<any> | Types.ObjectId | undefined | null): Types.ObjectId | null => {
  if (!target) {
    return null;
  }

  if (target instanceof Types.ObjectId) {
    return new Types.ObjectId(target.toString());
  }

  if (typeof target === 'object') {
    return target && target._id ? new Types.ObjectId(target._id) : null;
  }

  if (typeof target === 'string') {
    const result = fromGlobalId(target);

    if (result.type && result.id && Types.ObjectId.isValid(result.id)) {
      return new Types.ObjectId(result.id);
    }

    if (Types.ObjectId.isValid(target)) {
      return new Types.ObjectId(target);
    }

    return null;
  }

  return null;
};

type EnumObject = {
  [key: string]: string;
};

type EnumObjectResult = {
  [key: string]: {
    value: string;
  };
};

export const enumBuilderValues = (constants: EnumObject): EnumObjectResult =>
  Object.keys(constants).reduce(
    (prev, curr) => ({
      ...prev,
      [curr]: {
        value: constants[curr],
      },
    }),
    {},
  );
