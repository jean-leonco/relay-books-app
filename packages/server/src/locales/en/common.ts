const common = {
  SomethingWentWrong: 'Something went wrong.',
  TheDescriptionShouldNot: 'The description must be less than {{size}} characters.',
};
export default common;
export type Keys = keyof typeof common;
