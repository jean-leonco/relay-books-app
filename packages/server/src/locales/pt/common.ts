const common = {
  SomethingWentWrong: 'Algo deu errado.',
  TheDescriptionShouldNot: 'A descrição deve ter menos de {{size}} caracteres.',
};
export default common;
export type Keys = keyof typeof common;
