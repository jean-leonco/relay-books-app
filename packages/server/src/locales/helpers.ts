import i18next, { TOptions } from 'i18next';

import { NamespaceKeys, MessageKeys } from './en';

export const t = (namespace: NamespaceKeys, message: MessageKeys, options?: TOptions) =>
  i18next.t(`${namespace}::${message}`, options);
