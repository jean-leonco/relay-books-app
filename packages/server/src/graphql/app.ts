/* eslint-disable no-console */
import 'isomorphic-fetch';

import { koaPlayground } from 'graphql-playground-middleware';
import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import convert from 'koa-convert';
import cors from '@koa/cors';
import graphqlHttp from 'koa-graphql';
import koaLogger from 'koa-logger';
import Router from '@koa/router';
import idx from 'idx';
import i18next from 'i18next';
import koaI18next from 'koa-i18next';
import NoIntrospection from 'graphql-disable-introspection';
import { GraphQLError } from 'graphql';

import * as graphqlLoaders from '../loader';

import { JWT_KEY } from '../common/config';

import { KoaContextExt } from '../types';

import UserModel, { IUser } from '../modules/user/UserModel';

import { t } from '../locales/helpers';

import appPlatformMiddleware from '../modules/auth/appPlatformMiddleware';
import authMiddleware from '../modules/auth/authMiddleware';

import { getDataloaders } from './helper';
import { schema } from './schema';
import { i18nMiddleware } from './i18n';

const app = new Koa<any, KoaContextExt>();

if (process.env.NODE_ENV === 'production') {
  app.proxy = true;
}
app.keys = [JWT_KEY];

const loaders = graphqlLoaders;

const router = new Router<any, KoaContextExt>();

app.use(bodyParser());

app.on('error', (error) => {
  console.error('Error while answering request', { error });
});

if (process.env.NODE_ENV !== 'test') {
  app.use(koaLogger());
}

app.use(convert(cors({ maxAge: 86400, origin: '*' })));

router.get('/healthz', async (ctx) => {
  try {
    await UserModel.find().lean<IUser>();
    ctx.body = 'Database working';
    ctx.status = 200;
  } catch (err) {
    ctx.body = err.toString();
    ctx.status = 500;
  }
});

if (process.env.NODE_ENV !== 'production') {
  router.all(
    '/playground',
    koaPlayground({
      endpoint: '/graphql',
    }),
  );
}

// Middleware to get dataloaders
app.use(async (ctx, next) => {
  ctx.dataloaders = getDataloaders(loaders);
  await next();
});

app.use(
  koaI18next(i18next, {
    lookupCookie: 'lang', // detecting language in cookie
    order: ['header', 'cookie'],
    next: true,
  }),
);

app.use(i18nMiddleware);

// avoid requests from unknown clients
if (process.env.NODE_ENV !== 'development') {
  app.use(appPlatformMiddleware);
}

// Middleware to get result from authorization token
app.use(authMiddleware);

const graphqlSettingsPerReq = async (request, ctx, koaContext) => {
  const { dataloaders, user } = koaContext;
  const { appplatform, timezone } = request.header;

  if (process.env.NODE_ENV !== 'test') {
    console.info('Handling request', {
      appplatform,
      unauthorized: !user,
      userId: idx(user, (_) => _._id),
    });
  }

  return {
    graphiql: process.env.NODE_ENV !== 'production',
    schema,
    //rootValue: { request: ctx.req },
    validationRules: process.env.NODE_ENV === 'production' ? [NoIntrospection] : [],
    context: {
      user,
      dataloaders,
      appplatform,
      koaContext,
      timezone,
      t,
    },
    formatError: (error: GraphQLError) => {
      if (error.name && error.name === 'BadRequestError') {
        ctx.status = 400;
        ctx.body = 'Bad Request';
        return {
          message: 'Bad Request',
        };
      }

      if (error.path || error.name !== 'GraphQLError') {
        console.error(error);
      } else {
        console.log(`GraphQLWrongQuery: ${error.message}`);
      }

      console.error('GraphQL Error', { error });

      if (process.env.NODE_ENV !== 'production') {
        return {
          message: error.message,
          locations: error.locations,
          stack: error.stack,
        };
      } else {
        ctx.status = 400;
        ctx.body = 'Bad Request';
        return {
          message: 'Bad Request',
        };
      }
    },
  };
};

const graphqlServer = graphqlHttp(graphqlSettingsPerReq);
router.all('/graphql', graphqlServer);

app.use(router.routes()).use(router.allowedMethods());

export default app;
