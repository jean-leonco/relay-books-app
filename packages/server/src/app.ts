/* eslint-disable no-console */

import cors from '@koa/cors';
import Router from '@koa/router';
import { GraphQLError } from 'graphql';
import { koaPlayground } from 'graphql-playground-middleware';
import i18next from 'i18next';
import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import graphqlHttp from 'koa-graphql';
import koaI18next from 'koa-i18next';
import logger from 'koa-logger';

import { NODE_ENV, isProduction } from './config';

import { i18nMiddleware } from './i18n';
import { t } from './locales/helpers';
import appPlatformMiddleware from './modules/auth/appPlatformMiddleware';
import authMiddleware from './modules/auth/authMiddleware';
import { getDataloaders } from './modules/loader/loaderRegister';
import schema from './schema/schema';
import { KoaContext } from './types';

const app = new Koa<any, KoaContext>();
const router = new Router<any, KoaContext>();

app.use(bodyParser());
app.use(cors());

app.on('error', (error) => {
  console.error('Error while answering request', { error });
});

if (isProduction) {
  app.proxy = true;
}

if (NODE_ENV !== 'test') {
  app.use(logger());
}

if (!isProduction) {
  router.all('/playground', koaPlayground({ endpoint: '/graphql' }));
}

app.use(async (ctx, next) => {
  ctx.dataloaders = getDataloaders();
  await next();
});

app.use(
  koaI18next(i18next, {
    lookupCookie: 'lang',
    order: ['header', 'cookie'],
    next: true,
  }),
);

app.use(i18nMiddleware);

if (NODE_ENV !== 'development') {
  app.use(appPlatformMiddleware);
}

app.use(authMiddleware);

const graphqlSettingsPerReq = (request, ctx, koaContext) => {
  const { dataloaders, user } = koaContext;
  const { appplatform } = request.header;

  return {
    graphiql: !isProduction,
    schema,
    context: {
      user,
      dataloaders,
      koaContext,
      appplatform,
      t,
    },
    formatError: (error: GraphQLError) => {
      if (error.name && error.name === 'BadRequestError') {
        ctx.status = 400;
        ctx.body = 'Bad Request';
        return {
          message: 'Bad Request',
        };
      }

      if (error.path || error.name !== 'GraphQLError') {
        console.error(error);
      } else {
        console.log(`GraphQLWrongQuery: ${error.message}`);
      }

      console.error('GraphQL Error', { error });

      if (!isProduction) {
        return {
          message: error.message,
          locations: error.locations,
          stack: error.stack,
        };
      } else {
        ctx.status = 400;
        ctx.body = 'Bad Request';
        return {
          message: 'Bad Request',
        };
      }
    },
  };
};

const graphqlServer = graphqlHttp(graphqlSettingsPerReq);
router.all('/graphql', graphqlServer);

app.use(router.routes()).use(router.allowedMethods());

export default app;
