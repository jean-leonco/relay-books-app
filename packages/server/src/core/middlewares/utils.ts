import { GraphQLResolveInfo } from 'graphql';
import { ObjectSchema, Shape } from 'yup';

import { GraphQLContext } from '../../types';

interface GraphQLMutationConfig {
  validationSchema?(context: GraphQLContext): ObjectSchema<Shape<any, any>>;
  authenticatedOnly?: boolean;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): GraphQLMutationConfig | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
