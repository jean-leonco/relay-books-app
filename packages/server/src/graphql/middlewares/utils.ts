import { GraphQLField, GraphQLResolveInfo } from 'graphql';

import { Extensions } from '../../types';

interface Field extends GraphQLField<any, any> {
  extensions: Extensions | null | undefined;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): Field | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
