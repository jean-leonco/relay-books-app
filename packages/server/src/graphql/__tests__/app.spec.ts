import { httpRequestGraphql, sanitizeTestObject } from '@booksapp/test-utils';

import { clearDbAndRestartCounters, connectMongoose, disconnectMongoose, gql } from '../../../test/helpers';
import app from '../app';
import { PLATFORM } from '../../common/utils';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('App', () => {
  it('should return 200 even when token is null', async () => {
    const query = gql`
      query Q {
        me {
          id
          name
        }
      }
    `;

    const variables = {};
    const payload = { query, variables };

    const response = await httpRequestGraphql(payload, { appplatform: PLATFORM.APP, authorization: null }, app);

    expect(sanitizeTestObject(response.body)).toMatchSnapshot();
    expect(response.status).toBe(200);
  });
});
