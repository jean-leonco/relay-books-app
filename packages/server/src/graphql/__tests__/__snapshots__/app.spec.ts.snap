// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should return 200 even when token is null 1`] = `
Object {
  "data": Object {
    "me": "EMPTY",
  },
}
`;
